helix-core/src/main/java/org/apache/helix/HelixAdmin.java

```diff
@@ -367,6 +368,41 @@ void manuallyEnableMaintenanceMode(String clusterName, boolean enabled, String r
    */
   boolean isInMaintenanceMode(String clusterName);
 
+  /**
+   * Puts a cluster into pause mode, which will pause controller and participants.
+   * This can be used to retain the cluster state. When this method returns, it means
+   * the pause signal has been successfully sent, but it does not mean the cluster has
+   * fully entered pause mode. Because the cluster can take some time to finish the pause and
+   * process possible pending state transitions if necessary.
+   * <p>
+   * To check the cluster pause status, call {@link #getClusterPauseStatus(String)}.
+   *
+   * @param clusterName Participant cluster name.
+   * @param cancelPendingST whether or not cancel the pending state transitions for participants.
+   * @param reason The reason to put the cluster into pause mode.
+   */
+  void enableClusterPauseMode(String clusterName, boolean cancelPendingST, String reason);
+
+  /**
+   * Gets cluster pause status {@link ClusterPauseStatus}.
+   *
+   * @param clusterName cluster name.
+   * @return {@link ClusterPauseStatus}
+   */
+  ClusterPauseStatus getClusterPauseStatus(String clusterName);
+
+  /**
+   * Gets a cluster out of pause mode. When this method returns, it means
+   * the resume signal has been successfully sent, but it does not mean the cluster has
+   * fully exited pause mode.
+   * <p>
+   * To check whether the cluster has fully exited pause mode, call
+   * {@link #getClusterPauseStatus(String)}.
+   *
+   * @param clusterName Cluster name.
+   */
+  void disableClusterPauseMode(String clusterName);
```

Comment:
Why we need two APIs? enableClusterPauseMode can add an argument with enabled to be true or false. It follows our convention.

You can have private function to support enable/disable separately. But they can share the same API.

Reply:
I understand there are many similar APIs in Helix. However, to me, passing boolean parameters to differentiate the behaviors is really a code smell. Ref: https://softwareengineering.stackexchange.com/questions/208224/is-it-better-to-have-one-method-that-takes-a-bool-as-a-parameter-or-two-methods

I would try to avoid that. So I prefer two methods.